ocaml/xapi/xapi_vm.ml

```diff
@@ -246,22 +246,26 @@ let update_platform_secureboot ~__context ~self platform =
     platform
 
 let save_uefi_certificates_to_dir ~__context ~self =
-  let dir = try Sys.readdir !Xapi_globs.varstore_dir
-    with Sys_error _ ->
-      Xapi_stdext_unix.Unixext.mkdir_rec "/usr/share/varstored" 0o755;
-      debug "varstored directory does not exist"; [||] in
-
-  if not (Array.exists (fun e -> e = "PK.auth") dir) then
+  let uefi_platform_key = !Xapi_globs.varstore_dir ^ "PK.auth" in
+  if not (Sys.file_exists uefi_platform_key) then
     begin
-      let filename = "xapi_uefi_certificates.tar" in
       try
-        let file = Unix.openfile filename [Unix.O_RDWR; Unix.O_CREAT] 0o755 in
-        let contents = (Xapi_stdext_base64.Base64.decode (Db.Pool.get_uefi_certificates ~__context ~self)) in
-        Unixext.write_string_to_file filename contents;
-        Tar_unix.Archive.extract (fun _ -> !Xapi_globs.varstore_dir) file;
-        Unix.close file
-      with _ ->
-        debug "Error writing uefi certificates to file";
+        begin
+          let _dir = Sys.readdir !Xapi_globs.varstore_dir in
+          let filename = "xapi_uefi_certificates.tar" in
+          try
+            begin
+              let file = Unix.openfile filename [Unix.O_RDWR; Unix.O_CREAT] 0o755 in
+              let contents = (Xapi_stdext_base64.Base64.decode (Db.Pool.get_uefi_certificates ~__context ~self)) in
+              Unixext.write_string_to_file filename contents;
+              Tar_unix.Archive.extract (fun _ -> !Xapi_globs.varstore_dir) file;
+              finally (fun () -> Unix.close file) (fun () -> ())
+            end
+          with _ ->
```

Comment:
Would be good to know what the error was, there is a `D.log_and_ignore_exn` you could use for that.

Reply:
This with doesn't exist anymore by changing to if's above